### PR TITLE
Add Polymarket

### PR DESCRIPTION
**Pull request by tomfutago**
### Contribution type
Please check the type of contribution this pull request is for:

- [x] New spell(s)
- [ ] Adding to existing spell lineage
- [ ] Bug fix

---

### For new spell(s)
If you are building new spell(s), please provide the following information:

- **Spell name(s):**: ..
- **Description:**: nothing to see here, please disperse.. 

---

### CODE REVIEW

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql
Comment (0xBoxer) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:76]: now also in our db as `polymarket_polygon.FixedProductMarketMaker_evt_FPMMBuy`

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_capital_actions.sql
Comment (tomfutago) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_capital_actions.sql:33]: ok so how about:
- add `safe_proxies` spell - 2 key columns: `owner` and `proxy` - ideas for better names, plmk
- `capital_actions` will link back to the above spell (instead of hardcoded exchange addresses, as I'm doing currently)

> also koryo uses these 2 addresses, idk how to resolve that delta
> 
> 0x3A3BD7bb9528E159577F7C2e685CC81A765002E2
> 0x2791bca1f2de4661ed88a30c99a7a9449aa84174

not sure, what you mean by resolving delta for these addresses (WCOL & USDC contracts respectively)? I think with user in/out flows only USDC applies, no?

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql
Comment (tomfutago) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:89]: nope, it meant to be the same as either CTF `token0` or `token1` (for each there's cross join at token registration event, so I'm collapsing both to just one column expressed as `asset_id`). at least that's my line of thinking here..

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql
Comment (tomfutago) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:113]: hmm, looking at this now and I'd say current spell is actually correct.

so, in your [tx sample 0x878..](https://polygonscan.com/tx/0x878a300dce5517a50e0363668106cb770521d85dfa2b591e9d380c04028775f4) - there are 2 sides of the trade:
- 1 on Yes side (bought for $13.40) by [Fausto](https://polymarket.com/profile/0xcf2271dcd1803cc8f45cbfa0efa207cc78b4d46e?tab=activity)
- 1 on No side (bought for $6.60) by Mr 0xBoxer ;) 

different way to look at this: https://dune.com/queries/4031696

so again, I think it's all looking good, but if you have ideas how to represent this data in a better way - lmk

Comment (0xBoxer) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:113]: oh, that is my bad. I see that approach and it looks valid - however, the pricing still looks odd to me, do you have any idea what is happening there?

staying within this example, 0xBoxer bought 20 shares for 0.33 cents each,  for a total of $6.60. The current query spits out a price of `3.0303030303030303`. 
The same thing happens on the other side of the trade, with a price that's nowhere near close to the 0.67 it should be.

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql
Comment (0xBoxer) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:89]: I see, that makes sense to me.

So 2 AssetIds map to 1 conditionId?

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_capital_actions.sql
Comment (0xBoxer) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_capital_actions.sql]: > what you mean by resolving delta for these addresses 

lmao I fucked up with the USDC contract.

However, I think my point still stands, albeit in a slight variation:

The USDC only gets "locked up" in the conditional token contracts at 
```
 0x4D97DCd97eC945f40cF65F87097ACe5EA0476045
 0x3A3BD7bb9528E159577F7C2e685CC81A765002E2
 ``` 
 
 However, the market contracts might move them there, not the users directly so this will not be helpful while analysing user actions. 
 
 We could approximate Open Interest from these contracts, but not even that is clean, as there can be stale capital in there. that would be bets that are finished, but the winning tickets have not been redeemed yet.

Comment (0xBoxer) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_capital_actions.sql]: > add safe_proxies spell - 2 key columns: owner and proxy - ideas for better names, plmk
> capital_actions will link back to the above spell (instead of hardcoded exchange addresses, as I'm doing currently)

that sgtm

Comment (tomfutago) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_capital_actions.sql]: > The USDC only gets "locked up" in the conditional token contracts at
> 
> ```
>  0x4D97DCd97eC945f40cF65F87097ACe5EA0476045
>  0x3A3BD7bb9528E159577F7C2e685CC81A765002E2
> ```
> 
> However, the market contracts might move them there, not the users directly so this will not be helpful while analysing user actions.
> 
> We could approximate Open Interest from these contracts, but not even that is clean, as there can be stale capital in there. that would be bets that are finished, but the winning tickets have not been redeemed yet.

I'm still not clear on what needs to happen here. Is the aim to have both user actions and open interest all in one model? That's rather tricky, me thinks. How about:
- keep current `capital_actions` as it is, so it contains deposits to, withdrawals from and transfers in between proxy safes - so all USDC movements regardless of their source and destination within polymarket
- create a separate `open_interest` spell (eg based on: https://dune.com/queries/3343122)

will this approach capture what you have in mind or will something still be missing?

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql
Comment (tomfutago) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:89]: pretty much - in 99% cases multiple=2 assets really (Yes+No)

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql
Comment (tomfutago) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:113]: aah yep, fix is coming..

Comment (tomfutago) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_trades.sql:120]: should be good now (added `shares` column too): https://dune.com/queries/4031696

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_prices_daily.sql
Comment (0xRobin) [inline dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_market_prices_daily.sql:30]: this probably needs a `distinct`, same on the hourly model.

**Comment by 0xBoxer** [comment]
Due to a dbt version change, you need to merge main into your branch before the Github action runner runs successfully again.

**Review by 0xBoxer — APPROVED** [review]

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/_schema.yml
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/polymarket/polygon/_schema.yml:13]:
```suggestion
    data_tests:
```

Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/_schema.yml:13]: do a search/replace for all these instances. this was a change in the new version, where they introduced another test type. it's actually a warning on compile, but in CI, we force warnings to error out so we can try to avoid introducing new ones

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Comment by jeff-dude** [comment]
by you approving, does that mean you're ready / waiting for one of us to finalize and merge?

**Comment by 0xBoxer** [comment]
@jeff-dude if this run makes it, this should be good to merge, would appreciate if you can do that today still 😅 

should be fairly isolated and runtimes have not been crazy so far

**Comment by jeff-dude** [comment]
> @jeff-dude if this run makes it, this should be good to merge, would appreciate if you can do that today still 😅
> 
> should be fairly isolated and runtimes have not been crazy so far

happy to help if CI passes, but looks like things have broke?

**Comment by 0xBoxer** [comment]
we should be good now if I didn't screw up the incremental logic somehow

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_users.sql
Comment (0xRobin) [inline dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_users.sql]: we can use `min_by(.., block_time)` here to get the correct hash and address

**Comment by 0xBoxer** [comment]
okay now finally finally finally, this is all good to go @jeff-dude @0xRobin

File: dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_positions_raw.sql
Comment (0xBoxer) [inline dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_positions_raw.sql:7]: seeing that the incremental run runs for a long time, do we also switch this to `delete+insert`? 

also noticing that this doesn't have a partition keys specified, maybe that would help?

Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_positions_raw.sql:7]: i think this was isolated to a recent run, since it is running at the same time as your large solana PR. all PRs share the same cluster, so it's getting overloaded today.

you can see previous run today was more acceptable:
https://github.com/duneanalytics/spellbook/actions/runs/11553188411/job/32153823614

Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/polymarket/polygon/polymarket_polygon_positions_raw.sql]: with that in mind, i did still add partitions to the large raw tables at least

File: dbt_subprojects/daily_spellbook/models/_projects/polymarket/polygon/polymarket_polygon_positions_raw.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/polymarket/polygon/polymarket_polygon_positions_raw.sql:1]: i think this PR is all good with me. my only remaining question, prior to a merge:
should this model be using the new macro for balances?

it seems like the same type of pattern. a quick look at the daily_agg spell on top of raw balances has the similar `row_number()` function to get latest. there's a filter for specific token address. 

we may be missing an incremental filter on the raw balances that are left joined in here, as the only incremental filter appears to be on the sequence `days` CTE (which is fine for output, but lots of balances will be left joined).

let me know @0xRobin if you think it applies nicely to the balances macro, then we can look to push that final change here.

Comment (0xRobin) [reply dbt_subprojects/daily_spellbook/models/_projects/polymarket/polygon/polymarket_polygon_positions_raw.sql:1]: yes, we should look to apply the balances macro here.
We can't simply add the incremental filter on the raw balances as you might need balances arbitrarily in the past to perform the forward fill. 
But the macro takes care of that.

Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/_projects/polymarket/polygon/polymarket_polygon_positions_raw.sql:1]: yep, to clarify, i meant missing a filter relative to the macro. good callout.

i have attempted to apply the macro here now, let's see how it runs on first try 😅

Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/_projects/polymarket/polygon/polymarket_polygon_positions_raw.sql:1]: pretty smooth!

**Review by jeff-dude — APPROVED** [review]
gonna get this one in for the daily run tomorrow at midnight UTC

it all runs pretty efficiently, so any new PRs to fix should be straightforward 🙏

**Pull request merged**
